### PR TITLE
	[GEARPUMP-204]add unit test for external_hbase module

### PR DESCRIPTION
**Pull request by Roshanson**
[GEARPUMP-204]add unit test for external_hbase module

### CODE REVIEW

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 1]
```diff
@@ -19,39 +19,47 @@ package org.apache.gearpump.external.hbase
...
+class HBaseSink(userconfig: UserConfig, tableName: String, @transient var connection: Connection,
+    @transient var configuration: Configuration)
+  extends DataSink {
 
-class HBaseSink(
-    userconfig: UserConfig, tableName: String, @transient var configuration: Configuration)
-  extends DataSink{
-  lazy val connection = HBaseSink.getConnection(userconfig, configuration)
+
+  // var connection = HBaseSink.getConnection(userconfig, configuration)
```
Comment (huafengw) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: Why change the `connection`?

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 2]
```diff
@@ -87,19 +95,12 @@ class HBaseSink(
     connection.close()
   }
 
-  /**
-   * Overrides Java's default serialization
-   * Please do not remove this
-   */
```
Comment (huafengw) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: Why delete these doc?

**Comment by huafengw** [comment]
Generally my question is that is there any way implementing the HBaseSinkSpec without changing the HBaseSink?

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala (hunk 1 above)
Comment (Roshanson) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: I use  the parameter of  'connection'  to  separate real connection to HBase testing and unit testing.

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 3]
```diff
@@ -19,39 +19,47 @@ package org.apache.gearpump.external.hbase
...
   override def open(context: TaskContext): Unit = {}
 
   def this(userconfig: UserConfig, tableName: String) = {
-    this(userconfig, tableName, HBaseConfiguration.create())
+    this(userconfig, tableName, HBaseSink.getConnection(userconfig, HBaseConfiguration.create()),
+      HBaseConfiguration.create())
+  }
+  def this(userconfig: UserConfig, tableName: String, configuration: Configuration) = {
+    this(userconfig, tableName, HBaseSink.getConnection(userconfig, HBaseConfiguration.create()),
+      HBaseConfiguration.create())
```
Comment (manuzhang) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: `configuration` is passed in

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 4]
```diff
@@ -19,39 +19,47 @@ package org.apache.gearpump.external.hbase
...
   }
 
+
```
Comment (manuzhang) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: please remove these blank line changes

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 5]
```diff
@@ -115,15 +116,24 @@ object HBaseSink {
...
-    new HBaseSink(userconfig, tableName)
-  }
 
   def apply[T](userconfig: UserConfig, tableName: String, configuration: Configuration)
-    : HBaseSink = {
+  : HBaseSink = {
     new HBaseSink(userconfig, tableName, configuration)
   }
 
+  def apply[T](
```
Comment (manuzhang) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: this change is not necessary.

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 6]
```diff
@@ -148,11 +158,24 @@ object HBaseSink {
...
       ConnectionFactory.createConnection(configuration)
+
     } else {
       val user = UserProvider.instantiate(configuration)
         .create(UserGroupInformation.createRemoteUser(userName.get))
       ConnectionFactory.createConnection(configuration, user)
     }
 
+
+  }
```
Comment (manuzhang) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: need a blank line here

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala (hunk 1 above)
Comment (manuzhang) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: to allow it be mocked in test

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 7]
```diff
@@ -148,11 +158,24 @@ object HBaseSink {
...
+  def getConn(connection: Connection, isTest: Boolean, userConfig: UserConfig,
```
Comment (manuzhang) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: We don't need `isTest` here since this method is not supposed to be called in test

File: external/hbase/src/test/scala/org/apache/gearpump/external/hbase/HBaseSinkSpec.scala [hunk 8]
```diff
@@ -17,24 +17,52 @@
...
+    val userconfig = new UserConfig(map)
+    val tablename = "hbase"
+    val row = "row"
+    val group = "group"
+    val name = "name"
+    val value = "1.2"
+
+    when(conn.getTable(TableName.valueOf(tablename))).thenReturn(table)
+
+    val connection = HBaseSink.getConn(conn, true, userconfig, config)
```
Comment (manuzhang) [inline external/hbase/src/test/scala/org/apache/gearpump/external/hbase/HBaseSinkSpec.scala]: simply use mocked `conn` here

File: project/BuildExample.scala [hunk 9]
```diff
@@ -57,6 +57,7 @@ object BuildExample extends sbt.Build {
       )
   ) dependsOn(streaming % "test->test; provided", daemon % "test->test; provided")
 
+
```
Comment (manuzhang) [inline project/BuildExample.scala]: unnecessary changes

Comment (manuzhang) [inline project/BuildExample.scala]: unnecessary changes

**Comment by manuzhang** [comment]
@huafengw the current `HBaseSink` constructor is not testable. I'm afraid changing it is the only way to go.

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 10]
```diff
@@ -19,29 +19,33 @@ package org.apache.gearpump.external.hbase
...
 
   override def open(context: TaskContext): Unit = {}
 
   def this(userconfig: UserConfig, tableName: String) = {
-    this(userconfig, tableName, HBaseConfiguration.create())
+    this(userconfig, tableName, HBaseSink.getConnection(userconfig, HBaseConfiguration.create()),
+      HBaseConfiguration.create())
+  }
+  def this(userconfig: UserConfig, tableName: String, configuration: Configuration) = {
+    this(userconfig, tableName, HBaseSink.getConnection(userconfig, HBaseConfiguration.create()),
```
Comment (manuzhang) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: why create `configuration` here which is already passed in ?

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 11]
```diff
@@ -19,29 +19,33 @@ package org.apache.gearpump.external.hbase
...
-    userconfig: UserConfig, tableName: String, @transient var configuration: Configuration)
-  extends DataSink{
-  lazy val connection = HBaseSink.getConnection(userconfig, configuration)
   lazy val table = connection.getTable(TableName.valueOf(tableName))
 
   override def open(context: TaskContext): Unit = {}
 
   def this(userconfig: UserConfig, tableName: String) = {
-    this(userconfig, tableName, HBaseConfiguration.create())
+    this(userconfig, tableName, HBaseSink.getConnection(userconfig, HBaseConfiguration.create()),
```
Comment (manuzhang) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: why create configuration twice here ?

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 12]
```diff
@@ -97,7 +101,7 @@ class HBaseSink(
   }
 
   /**
-   * Overrides Java's default deserialization
+   * Overrides Java's default serialization
```
Comment (manuzhang) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: serialization ?

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 13]
```diff
@@ -114,16 +118,19 @@ object HBaseSink {
   val COLUMN_FAMILY = "hbase.table.column.family"
   val COLUMN_NAME = "hbase.table.column.name"
   val HBASE_USER = "hbase.user"
-
-  def apply[T](userconfig: UserConfig, tableName: String): HBaseSink = {
-    new HBaseSink(userconfig, tableName)
-  }
+  private val CONFIG = HBaseConfiguration.create()
```
Comment (manuzhang) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: will configuration be created at client side ?

**Comment by manuzhang** [comment]
@huafengw any more comments ?

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala (hunk 5 above)
Comment (huafengw) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: Code format

**Comment by huafengw** [comment]
One question, now the `connection` is a transient value, will there be a NPE after the HBaseSink deserialized?

File: external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala [hunk 14]
```diff
@@ -19,27 +19,33 @@ package org.apache.gearpump.external.hbase
...
-  extends DataSink{
-  lazy val connection = HBaseSink.getConnection(userconfig, configuration)
+  lazy val connection = conn(userconfig, configuration)
   lazy val table = connection.getTable(TableName.valueOf(tableName))
 
   override def open(context: TaskContext): Unit = {}
 
+  def this(userconfig: UserConfig, tableName: String, configuration: Configuration) = {
+    this(userconfig, tableName, (userconfig: UserConfig, config: Configuration) =>
+      {HBaseSink.getConnection(userconfig, config)}, configuration)
```
Comment (manuzhang) [inline external/hbase/src/main/scala/org/apache/gearpump/external/hbase/HBaseSink.scala]: just passing in `getConnection` will do, I think

**Comment by huafengw** [comment]
+1
